src/main/java/teammates/storage/sqlapi/UserDb.java

```diff
@@ -0,0 +1,171 @@
+package teammates.storage.sqlapi;
+
+import java.time.Instant;
+
+import org.hibernate.Session;
+
+import teammates.common.exception.EntityAlreadyExistsException;
+import teammates.common.exception.EntityDoesNotExistException;
+import teammates.common.exception.InvalidParametersException;
+import teammates.common.util.HibernateUtil;
+import teammates.storage.sqlentity.Instructor;
+import teammates.storage.sqlentity.Student;
+import teammates.storage.sqlentity.User;
+
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Root;
+
+/**
+ * Handles CRUD operations for users.
+ *
+ * @see User
+ */
+public final class UserDb extends EntitiesDb<User> {
+    private static final UserDb instance = new UserDb();
+
+    private UserDb() {
+        // prevent initialization
+    }
+
+    public static UserDb inst() {
+        return instance;
+    }
+
+    /**
+     * Creates a user.
+     */
+    public <T extends User> T createUser(T user)
+            throws InvalidParametersException, EntityAlreadyExistsException {
+        assert user != null;
+
+        if (!user.isValid()) {
+            throw new InvalidParametersException(user.getInvalidityInfo());
+        }
+
+        String courseId = user.getCourse().getId();
+        String email = user.getEmail();
+
+        if (doesUserExist(courseId, email)) {
```

Comment:
I see the issue here, will need to split this method back into instructor/student. Validation for both would be different, since course/email is not unique for user.

Reply:
Thanks for your comment samuel! By splitting back into instructor/student does that mean reverting back to something like this?

```java
        // This logic is definitely wrong 😓 
        if (getInstructor(user.getId()) == null) {
            throw new EntityAlreadyExistsException(String.format(ERROR_CREATE_ENTITY_ALREADY_EXISTS, user.toString()));
        }

        if (getStudent(user.getId()) == null) {
            throw new EntityAlreadyExistsException(String.format(ERROR_CREATE_ENTITY_ALREADY_EXISTS, user.toString()));
        }
```

I checked on the EntitiesDb hasExisting methods in both Instructor and Student:
**Instructor**
![image](https://user-images.githubusercontent.com/46486515/221078743-2b6213a0-e737-4bf4-9604-75f2fe0fda93.png)

`generateId` in Instructor
![image](https://user-images.githubusercontent.com/46486515/221078995-ea97d2a5-ef22-472c-b0a0-4ed35f67a18f.png)

**Student**
![image](https://user-images.githubusercontent.com/46486515/221078807-e3086763-d6cc-4732-927e-2d46e2be1588.png)

`generateId` in Student
![image](https://user-images.githubusercontent.com/46486515/221078943-93a6ba08-875c-4317-b3f7-085431553d79.png)

Do I follow these above? If I am, may I ask how should I go about doing it for `Student`? Noted that course + email is not unique for user too, for this new implementation it would be `id` itself?

Reply:
No, split createUser into createInstructor and createStudent. Validate existence separately for both.

Reply:
Okay, thanks samuel!!

Reply:
One last thing, you mentioned that validation will be different for both, are we not using courseId and email anymore? How should I do the validation? A little lost on this part

Reply:
Same, but need hasExistngInstructor and hasExistingStudent. courseId and email is only unique for instructor/student, not for users in general.

Reply:
Resolved! Thanks samuel! Hope it's good now!